### PR TITLE
Add sparse initialization

### PR DESCRIPTION
**Pull request by atiyo**
Add sparse initialization, documentation and tests. Trim whitespace in editted files.

This PR is intended to address one of the outstanding points in bringing Flux to parity with PyTorch's features so it partially addresses https://github.com/FluxML/Flux.jl/issues/1431 and fully addresses https://github.com/FluxML/Flux.jl/issues/1450.

The implementation follows the method given in [PyTorch implementation](https://pytorch.org/docs/stable/_modules/torch/nn/init.html#sparse_): a normally-distributed array is created, then a fixed proportion of randomly chosen row-indices is zeroed out for every column. Like the PyTorch version, it is restricted to 2-d Arrays.

### PR Checklist

- [x] Tests are added
- [x] Entry in NEWS.md
- [x] Documentation, if applicable
- [ ] Final review from `@dhairyagandhi96` (for API changes).

### CODE REVIEW

File: src/utils.jl [hunk 1]
```diff
@@ -170,14 +174,63 @@ end
 kaiming_normal(dims...; kwargs...) = kaiming_normal(Random.GLOBAL_RNG, dims...; kwargs...)
 kaiming_normal(rng::AbstractRNG; kwargs...) = (dims...; kwargs...) -> kaiming_normal(rng, dims...; kwargs...)
 
+"""
+    sparse([rng=GLOBAL_RNG], dims...; sparsity, std = 0.01)
```
Comment (atiyo) [inline src/utils.jl:178]: Is calling this function `sparse` ok? It seems like a very general name to be taken. This keeps consistency with the other initialization names, though.

Also, I couldn't find a specific naming convention for standard deviations as keyword arguments. Using \sigma could be an alternative, but is often used in Flux to denote activations, so I preferred to avoid it.

File: src/utils.jl [hunk 2]
```diff
@@ -170,14 +174,63 @@ end
...
+  if length(dims) != 2
+    throw(ArgumentError("Only 2-dimensional outputs are supported for sparse initialization."))
+  end
+  rows, cols = dims
+  prop_zero = min(1.0, sparsity)
+  num_zeros = ceil(Integer, prop_zero * rows)
+  sparse_array = randn(rng, Float32, dims...) .* Float32(std)
+  row_indices = collect(1:rows)
+  for col in 1:cols
+    randperm!(row_indices)
```
Comment (atiyo) [inline src/utils.jl:216]: Using `randperm!` and then index some elements as a shortcut to sampling without replacement.

An alternative would be to use `sample` from `StatsBase`, which is already a dependency, but not used outside of the tests. I preferred to use `randperm!` from `Random` in order to avoid adding this extra import.

File: src/utils.jl (hunk 1 above)
Comment (DhairyaLGandhi) [reply src/utils.jl:178]: This would cause banking conflicts with Sparse arrays, so I would pick a different name.

File: src/utils.jl [hunk 3]
```diff
@@ -170,14 +174,63 @@ end
...
+
+[1] Martens, J, "Deep learning via Hessian-free optimization" _Proceedings of the 27th International Conference on International Conference on Machine Learning_. 2010.
+"""
+function sparse(rng::AbstractRNG, dims...; sparsity, std = 0.01)
+  if length(dims) != 2
+    throw(ArgumentError("Only 2-dimensional outputs are supported for sparse initialization."))
+  end
+  rows, cols = dims
+  prop_zero = min(1.0, sparsity)
+  num_zeros = ceil(Integer, prop_zero * rows)
```
Comment (DhairyaLGandhi) [inline src/utils.jl:212]: Use `\div` here

File: src/utils.jl (hunk 2 above)
Comment (DhairyaLGandhi) [reply src/utils.jl:216]: Seems like we should be able to generate a mask and multiply too ratherv then iterating over the columns.

File: src/utils.jl [hunk 4]
```diff
@@ -170,14 +174,63 @@ end
...
+  sparse_array = randn(rng, Float32, dims...) .* Float32(std)
+  row_indices = collect(1:rows)
+  for col in 1:cols
+    randperm!(row_indices)
+    zero_indices = row_indices[1:num_zeros]
+    sparse_array[zero_indices, col] .= 0.0
```
Comment (DhairyaLGandhi) [inline src/utils.jl:218]: Type instability possibility. If we can take the type as input, we can use `zero(T)`

**Comment by DhairyaLGandhi** [comment]
Thanks for looking into this! I've left a couple of thoughts in the implementation. We would need to use a different name though since `sparse` is already a function in a stdlib

File: src/utils.jl (hunk 1 above)
Comment (CarloLucibello) [reply src/utils.jl:178]: what about `sparse_init`?

File: src/utils.jl (hunk 4 above)
Comment (CarloLucibello) [inline src/utils.jl:218]:
```suggestion
    sparse_array[zero_indices, col] .= 0
```
There is no risk of instability in any case, we are assigning in-place

File: src/utils.jl (hunk 1 above)
Comment (atiyo) [reply src/utils.jl]: My first thought was also `sparse_init`, so I've run with that.

File: src/utils.jl (hunk 2 above)
Comment (atiyo) [reply src/utils.jl]: Agreed that iterating over columns was a bit unwieldy. Manually setting zeros then shuffling rows with `mapslice(shuffle,...)` was the cleanest I could come up with.

File: src/utils.jl (hunk 3 above)
Comment (atiyo) [reply src/utils.jl:212]: I assume you mean something like `div(rows, 1/prop_zero)`? This returns a float since `prop_zero` is a float, so would require further casting to an integer. I thought above was a bit easier to follow, but am happy to go with what you think is best.

File: src/utils.jl (hunk 4 above)
Comment (atiyo) [reply src/utils.jl]: In the corresponding refactor, I use `zero(Float32)` when assigning zeros. Happy to take the type as an input and pipe it through. But this isn't the case with other initialization schemes, and I was trying to keep things as consistent as possible.

Comment (atiyo) [reply src/utils.jl]: Appreciate the suggestion. In the refactor I assign to `zeros(Float32)`, which I appreciate might be needlessly type specific. Happy to change it to an assignment to `0` based on consensus.

File: src/utils.jl [hunk 5]
```diff
@@ -170,14 +174,58 @@ end
...
+"""
+function sparse_init(rng::AbstractRNG, dims...; sparsity, std = 0.01)
+  if length(dims) != 2
+    throw(ArgumentError("Only 2-dimensional outputs are supported for sparse initialization."))
+  end
+  rows, cols = dims
+  prop_zero = min(1.0, sparsity)
+  num_zeros = ceil(Integer, prop_zero * rows)
+  sparse_array = randn(rng, Float32, dims...) .* Float32(std)
+  sparse_array[1:num_zeros, :] .= zero(Float32)
```
Comment (ToucheSir) [inline src/utils.jl:214]: You can use Float32 constants instead of casting or calling `zero`.
```suggestion
  sparse_array[1:num_zeros, :] .= 0f0
```

**Comment by CarloLucibello** [comment]
Current implementation does `  sparse_array[1:num_zeros, :] .= 0`, but it should randomize the zero positions in each row.
The pytorch code is:
```python
def sparse_(tensor, sparsity, std=0.01):
    r"""Fills the 2D input `Tensor` as a sparse matrix, where the
    non-zero elements will be drawn from the normal distribution
    :math:`\mathcal{N}(0, 0.01)`, as described in `Deep learning via
    Hessian-free optimization` - Martens, J. (2010).

    Args:
        tensor: an n-dimensional `torch.Tensor`
        sparsity: The fraction of elements in each column to be set to zero
        std: the standard deviation of the normal distribution used to generate
            the non-zero values

    Examples:
        >>> w = torch.empty(3, 5)
        >>> nn.init.sparse_(w, sparsity=0.1)
    """
    if tensor.ndimension() != 2:
        raise ValueError("Only tensors with 2 dimensions are supported")

    rows, cols = tensor.shape
    num_zeros = int(math.ceil(sparsity * rows))

    with torch.no_grad():
        tensor.normal_(0, std)
        for col_idx in range(cols):
            row_indices = torch.randperm(rows)
            zero_indices = row_indices[:num_zeros]
            tensor[zero_indices, col_idx] = 0
    return tensor
```
We should follow them, swapping cols with rows, so `sparse_array[:, zero_indices] .= 0`

**Comment by CarloLucibello** [comment]
Sorry, now I see that you randomly permute with `mapslices(shuffle, x)`. So only my comment of swapping rows and cols compared to python's implementation applies

**Comment by atiyo** [comment]
> Sorry, now I see that you randomly permute with `mapslices(shuffle, x)`. So only my comment on swapping rows and cols compared to python's implementation applies

I'm not clear on why we need to swap rows and cols compared to PyTorch. I understand the batch dimension is different, but as far as I could tell Flux uses similar shapes for weights.

E.g
```python
In [1]: from torch import nn
In [2]: nn.Linear(1,2).weight.shape
Out[2]: torch.Size([2, 1])
```
```julia
julia> using Flux
julia> size(Dense(1,2).W)
(2, 1)
```

**Comment by CarloLucibello** [comment]
@atiyo you're right, I always thought that pytorch applies the transform `x * W`,  but now I see that it does `x * W^T` instead. So this PR looks entirely fine to me. If we want to do some more name bikeshedding, `init_sparse` is an alternative proposal maybe more discoverable by tab-completion when looking for initialization methods.

Another consideration is that maybe we can move initialization functions to a submodule, but this doesn't have to be necessarily discussed here.

File: src/utils.jl (hunk 3 above)
Comment (DhairyaLGandhi) [reply src/utils.jl:212]: `\div{tab}` should return an inyt

**Comment by DhairyaLGandhi** [comment]
Let's not move it to a submodule, doesn't seem worthwhile enough as a standalone to me

`init_sparse` would be better for discoverability, but somewhat inconsistent. I guess it's alright for now

File: src/utils.jl (hunk 3 above)
Comment (atiyo) [reply src/utils.jl:212]: I might be missing something. I'm finding `÷` to behave the same way as `div`, i.e. it's returning a float for float values of `prop_zero`.

```julia
julia> prop_zero = 0.11; rows = 50;
julia> ÷(rows, 1/prop_zero, RoundUp)
6.0
```

Using `÷` as an infix operator still returns a float, but also doesn't allow to specify a `RoundingMode`. We need to round up to maintain consistency with PyTorch.

**Review by CarloLucibello — APPROVED** [review]
